maserol/preprocess.py

```diff
@@ -48,45 +54,51 @@ def get_affinity(receptor, abs):
-        num = "I"
-    elif (match == '2' or match.lower() == 'ii'):
-        num = "II"
-    else:
-        num = "III"
-
-    # search for receptor match in affinities dataArray
-    for r in list(df.index):
-        r_regex = "fc[gr]*" + num + receptor[x.end()::]
-        if re.match(r_regex, r, flags=re.IGNORECASE):
-            return df.at[r,abs]
-    return 0
-
-
-def assembleKav(data: xr.DataArray, fucose=True):
+    if x is not None: 
+        match = x.group()
+        if (match == '1' or match.lower() == 'i'):
+            num = "I"
+        elif (match == '2' or match.lower() == 'ii'):
+            num = "II"
+        else:
+            num = "III"
+
+        # search for receptor match in affinities dataArray
+        for r in list(df.index):
+            r_regex = "fc[gr]*" + num + receptor[x.end()::]
+            if re.match(r_regex, r, flags=re.IGNORECASE):
+                return df.at[r,ab_type]
+    # the receptor was not an FCR
+    raise AffinityNotFoundException(receptor, ab_type)
+
+
+def assembleKav(data: xr.DataArray, ab_types: Optional[Iterable]=DEFAULT_AB_TYPES) -> xr.DataArray:
     """ Assemble affinity matrix for a given dataset. """
-    absf = HIgGs if not fucose else HIgGFs
     receptors = data.Receptor.values    # work even when data did not go thru prepare_data()
-    igg = [x for x in receptors if (re.match("^igg", x, flags=re.IGNORECASE))]
-    fc = [x for x in receptors if (re.match("fc[gr]*", x, flags=re.IGNORECASE) and x != "FcRalpha")]
+    igg_receptors = [x for x in receptors if (re.match("^igg", x, flags=re.IGNORECASE))]
+    other_receptors = [x for x in receptors if x not in igg_receptors]
+    # fc = [x for x in receptors if (re.match("fc[gr]*", x, flags=re.IGNORECASE) and x != "FcRalpha")]
+    # other_receptors = [x for x in receptors if not (x in igg or x in fc)] # Includes complement
 
     # assemble matrix
-    Kav = xr.DataArray(np.full((len(igg+fc), len(absf)), 10),
-                       coords=[igg+fc, absf],
+    Kav = xr.DataArray(np.full((len(igg_receptors+other_receptors), len(ab_types)), 10),
+                       coords=[igg_receptors+other_receptors, list(ab_types)],
                        dims=["Receptor", "Abs"])
 
     # fill in all IgG - IgG pair affinity values
-    for ab in absf:
-        for ig in igg:
+    for ab in ab_types:
+        for ig in igg_receptors:
             if (ab == ig or ab[:-1] == ig):
                 Kav.loc[dict(Receptor=ig, Abs=ab)] = 10**8      # default affinity for anti-IgGx Ab
 
     # fill in remaining affinity values
-    for ab in absf:
-        for r in fc:
+    for ab in ab_types:
+        for r in other_receptors:
             if not (ab[-1]=='f' and re.match("fc[gr]*3", r, flags=re.IGNORECASE)):
-                affinity = get_affinity(r, ab[:4])
+                try:
+                    affinity = get_affinity(r, ab[:4])
+                except AffinityNotFoundException:
+                    affinity = np.random.uniform(1E5, 5E5)
```

Comment:
Why does it fill in a random value if the affinity is not found? It shouldn't do that. The `prepare_data` step should remove all unreasonable receptors so that everything in this matrix can be looked up.